python/ray/autoscaler/v2/instance_manager/subscribers/instance_launcher.py

```diff
@@ -2,159 +2,246 @@
-            AUTOSCALER_MAX_CONCURRENT_LAUNCHES / float(AUTOSCALER_MAX_LAUNCH_BATCH)
-        ),
-        max_instances_per_request: int = AUTOSCALER_MAX_LAUNCH_BATCH,
+        node_provider: ICloudInstanceProvider,
+        upscaling_speed: float,
+        max_launch_per_type: int = AUTOSCALER_MAX_LAUNCH_PER_TYPE,
+        launch_callback: Optional[Callable] = None,
     ) -> None:
         self._instance_storage = instance_storage
         self._node_provider = node_provider
-        self._max_concurrent_requests = max_concurrent_requests
-        self._max_instances_per_request = max_instances_per_request
         self._executor = ThreadPoolExecutor(max_workers=1)
-        self._launch_instance_executor = ThreadPoolExecutor(
-            max_workers=self._max_concurrent_requests
-        )
+        self._upscaling_speed = upscaling_speed
+        self._max_launch_per_type = max_launch_per_type
+        self._launch_callback = launch_callback or self._default_launch_callback
+        self._request_prefix = "instance_launcher"
 
     def notify(self, events: List[InstanceUpdateEvent]) -> None:
-        # TODO: we should do reconciliation based on events.
-        has_new_request = any(
-            [event.new_instance_status == Instance.UNKNOWN for event in events]
+        has_new_queued = any(
+            [event.new_instance_status == Instance.QUEUED for event in events]
         )
-        if has_new_request:
-            self._executor.submit(self._may_launch_new_instances)
 
-    def _may_launch_new_instances(self):
-        new_instances, _ = self._instance_storage.get_instances(
-            status_filter={Instance.UNKNOWN}
+        if has_new_queued:
+            fut = self._executor.submit(self._may_launch_new_instances)
+            fut.add_done_callback(self._launch_callback)
+
+    ############################
+    # Private functions
+    ############################
+
+    @staticmethod
+    def _default_launch_callback(fut):
+        """
+        Default callback for launching instances.
+
+        One could modify this for testing by passing in a different callback
+        when initializing the instance launcher.
+
+        Args:
+            fut: the future of the launch request.
+        """
+        try:
+            ret = fut.result()
+        except Exception as e:
+            logger.exception(f"Failed to request instances: {str(e)}")
+        else:
+            logger.info(f"Successfully request instances to launch: {dict(ret)}")
+
+    def _may_launch_new_instances(self) -> Optional[Dict[str, Dict[str, int]]]:
+        """
+        Launches new instances if there are new instances in QUEUED status.
+
+        In order to avoid launching too many instances at the same time, we
+        limit the number of REQUESTED per node type as follows:
+        - Limit the total number of REQUESTED instances to be
+            max(self._max_launch_per_type, self._upscaling_speed * num_requested_or_allocated) # noqa
```

Comment:
This is the current behaviour: 

https://github.com/ray-project/ray/blob/52f7b898659b3a211575cd397ce8967cededd4f3/doc/source/cluster/vms/user-guides/configuring-autoscaling.rst#L47-L48